Building for ARM fails

With the latest EH changes the project fails building on both 32- and 64-bit ARM.

**armeabi-v7a**
```
eh_personality.c:571:22: error: no member named 'private_1' in 'struct _Unwind_Exception'
                ex->cxx_exception->private_1 = exceptionObject->private_1;
                ~~~~~~~~~~~~~~~~~  ^
eh_personality.c:571:51: error: no member named 'private_1' in 'struct _Unwind_Exception'
                ex->cxx_exception->private_1 = exceptionObject->private_1;
                                               ~~~~~~~~~~~~~~~  ^
eh_personality.c:572:22: error: no member named 'private_2' in 'struct _Unwind_Exception'
                ex->cxx_exception->private_2 = exceptionObject->private_2;
                ~~~~~~~~~~~~~~~~~  ^
eh_personality.c:572:51: error: no member named 'private_2' in 'struct _Unwind_Exception'
                ex->cxx_exception->private_2 = exceptionObject->private_2;
                                               ~~~~~~~~~~~~~~~  ^
eh_personality.c:576:20: error: no member named 'private_1' in 'struct _Unwind_Exception'
                exceptionObject->private_1 = ex->cxx_exception->private_1;
                ~~~~~~~~~~~~~~~  ^
eh_personality.c:576:51: error: no member named 'private_1' in 'struct _Unwind_Exception'
                exceptionObject->private_1 = ex->cxx_exception->private_1;
                                             ~~~~~~~~~~~~~~~~~  ^
eh_personality.c:577:20: error: no member named 'private_2' in 'struct _Unwind_Exception'
                exceptionObject->private_2 = ex->cxx_exception->private_2;
                ~~~~~~~~~~~~~~~  ^
eh_personality.c:577:51: error: no member named 'private_2' in 'struct _Unwind_Exception'
                exceptionObject->private_2 = ex->cxx_exception->private_2;
                                             ~~~~~~~~~~~~~~~~~  ^
```

**arm64-v8a**
```
CMakeFiles/objc.dir/eh_personality.c.o: In function `internal_objc_personality':
eh_personality.c:392: undefined reference to `cxx_exception_class'
eh_personality.c:392: undefined reference to `cxx_exception_class'
eh_personality.c:394: undefined reference to `test_cxx_eh_implementation'
CMakeFiles/objc.dir/eh_personality.c.o: In function `__gnustep_objcxx_personality_v0':
eh_personality.c:558: undefined reference to `cxx_exception_class'
eh_personality.c:558: undefined reference to `cxx_exception_class'
eh_personality.c:560: undefined reference to `test_cxx_eh_implementation'
CMakeFiles/objc.dir/eh_personality.c.o: In function `objc_begin_catch':
eh_personality.c:627: undefined reference to `cxx_exception_class'
eh_personality.c:627: undefined reference to `cxx_exception_class'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

As a side-note, it’d be great to have CI for ARM as well. Cirrus CI doesn’t support this yet, but [hopefully soon](https://github.com/cirruslabs/cirrus-ci-docs/issues/218#issuecomment-632720357). I couldn’t figure out whether Azure supports this.

> These are supposed to just be copying the state, we should be able to add a `COPY_EXCEPTION` macro in the two unwind headers to do the right thing here.

I’ve given this a try in #163. Let me know if there’s anything I should change.

I realized that the errors I was seeing on arm64 (undefined reference to cxx_exception_class etc.) were caused by building without a usable C++ runtime. Not sure if that should be a supported use case (I wouldn’t need it) – if not maybe "No useable C++ runtime found" should be an error.

> For other projects, we're using the Docker + QEMU thing on Azure, which should be fine for this.

Good point. I realized it’d be great to specifically test building for Android, which should work fine on CI using cross-compilation. I’ll look into this once libc++abi support lands.

--- a/eh_personality.c
+++ b/eh_personality.c
@@ -568,13 +568,11 @@ BEGIN_PERSONALITY_FUNCTION(__gnustep_objcxx_personality_v0)
 		}
 		// We now have two copies of the _Unwind_Exception object (which stores
 		// state for the unwinder) in flight.  Make sure that they're in sync.
-		ex->cxx_exception->private_1 = exceptionObject->private_1;
-		ex->cxx_exception->private_2 = exceptionObject->private_2;
+		COPY_EXCEPTION(ex->cxx_exception, exceptionObject)
 		exceptionObject = ex->cxx_exception;
 		exceptionClass = cxx_exception_class;
 		int ret = CALL_PERSONALITY_FUNCTION(__gxx_personality_v0);
-		exceptionObject->private_1 = ex->cxx_exception->private_1;
-		exceptionObject->private_2 = ex->cxx_exception->private_2;
+		COPY_EXCEPTION(exceptionObject, ex->cxx_exception)
 		if (ret == _URC_INSTALL_CONTEXT)
 		{
 			get_thread_data()->cxxCaughtException = YES;

--- a/unwind-arm.h
+++ b/unwind-arm.h
@@ -201,3 +201,9 @@ _Unwind_Reason_Code name(_Unwind_State state,\
 	_Unwind_SetGR (context, 12, (unsigned long)exceptionObject);
 
 #define CALL_PERSONALITY_FUNCTION(name) name(state,exceptionObject,context)
+
+#define COPY_EXCEPTION(dst, src) \
+  memcpy((dst)->unwinder_cache, (src)->unwinder_cache, sizeof(struct unwinder_cache)); \
+  memcpy((dst)->barrier_cache, (src)->barrier_cache, sizeof(struct barrier_cache)); \
+  memcpy((dst)->cleanup_cache, (src)->cleanup_cache, sizeof(struct cleanup_cache)); \
+  memcpy((dst)->pr_cache, (src)->pr_cache, sizeof(struct pr_cache));

--- a/unwind-itanium.h
+++ b/unwind-itanium.h
@@ -164,6 +164,10 @@ _Unwind_Reason_Code name(int version,\
 
 #define CALL_PERSONALITY_FUNCTION(name) name(version, actions, exceptionClass, exceptionObject, context)
 
+#define COPY_EXCEPTION(dst, src) \
+  (dst)->private_1 = (src)->private_1; \
+  (dst)->private_2 = (src)->private_2;
+
 #ifdef __cplusplus
 }
 #endif

